lib/azure/armrest/configuration.rb

```diff
@@ -0,0 +1,256 @@
+
+      # The client key used to gather token information.
+      attr_accessor :client_key
+
+      # The tenant ID used to gather token information.
+      attr_accessor :tenant_id
+
+      # The subscription ID used for each http request.
+      attr_accessor :subscription_id
+
+      # The resource group used for http requests.
+      attr_accessor :resource_group
+
+      # The grant type. The default is client_credentials.
+      attr_accessor :grant_type
+
+      # The content type specified for http requests. The default is 'application/json'
+      attr_accessor :content_type
+
+      # The accept type specified for http request results. The default is 'application/json'
+      attr_accessor :accept
+
+      # Explicitly set the token.
+      attr_writer :token
+
+      # Proxy to be used for all http requests.
+      attr_accessor :proxy
+
+      # SSL version to be used for all http requests.
+      attr_accessor :ssl_version
+
+      # SSL verify mode for all http requests.
+      attr_accessor :ssl_verify
+
+      # Namespace providers, their resource types, locations and supported api-version strings.
+      attr_reader :providers
+
+      # Yields a new Azure::Armrest::Configuration objects. Note that you must
+      # specify a client_id, client_key, tenant_id and subscription_id. All other
+      # parameters are optional.
+      #
+      # Example:
+      #
+      #   config = Azure::Armrest::Configuration.new(
+      #     :client_id       => 'xxxx',
+      #     :client_key      => 'yyyy',
+      #     :tenant_id       => 'zzzz',
+      #     :subscription_id => 'abcd'
+      #   )
+      #
+      # If you specify a :resource_group, that group will be used for resource
+      # group based service class requests. Otherwise, you will need to specify
+      # a resource group for most service methods.
+      #
+      # Although you can specify an :api_version, it is typically overridden
+      # by individual service classes.
+      #
+      def initialize(args)
+        # Use defaults, and override with provided arguments
+        options = {
+          :api_version      => '2015-01-01',
+          :accept           => 'application/json',
+          :content_type     => 'application/json',
+          :grant_type       => 'client_credentials',
+          :token_expiration => Time.new(0).utc,
+          :proxy            => ENV['http_proxy'],
+          :ssl_version      => 'TLSv1',
+        }.merge(args)
+
+        options.each { |key, value| send("#{key}=", value) }
+
+        unless client_id && client_key && tenant_id && subscription_id
+          raise ArgumentError, "client_id, client_key, tenant_id and subscription_id must all be specified"
+        end
+
+        # Allows for URI objects or Strings.
+        @proxy = @proxy.to_s if @proxy
+
+        # A one-time lookup table set at the class level
+        @providers = fetch_providers
+        set_provider_info(@providers) if self.class.provider_cache.empty?
```

Comment:
Seems a problem since `@providers` is per instance but `provider_cache` is per class while it is populated only once. Looks like `@providers` should be per class also.